File: problem-2/2.cpp
#include <cstdio>

using namespace std;

int memoization[ 10000 ];

int fib( int n ) {
    if ( n == 0 || n == 1 ) {
        return 1;
    }
    if ( memoization[ n ] == 0 ) {
        memoization[ n ] = fib( n - 1 ) + fib( n - 2 );
    }
    return memoization[ n ];
}

int main() {
    int f = fib( 0 );
    int i = 0, sum = 0;

    while ( f <= 4000000 ) {
        f = fib( i );
        if ( f % 2 == 0 ) {
            sum += f;
        }
        ++i;
    }
    printf( "%i\n", sum );

    return 0;
}

File: problem-2/2.py
from itertools import takewhile

N = 4000000

def fib():
    a, b = 0, 1
    while True:
        a, b = b, a + b
        yield b

seq = fib()
seq = takewhile(lambda x: x <= N, seq)
seq = filter(lambda x: x % 2 == 0, seq)

print sum(seq)
